src/components/Icon/Icon.tsx

```diff
@@ -15,12 +16,49 @@ const COLORS_WITH_BACKDROPS = [
   'inkLighter',
 ];
 
+const NEW_DESIGN_LANGUAGE_COLORS = [
+  'base',
+  'disabled',
+  'hovered',
+  'pressed',
+  'subdued',
+  'critical',
+  'warning',
+  'highlight',
+  'interactive',
+  'success',
+  'primary',
+  'primaryDisabled',
+  'primaryHovered',
+  'primaryPressed',
+];
+
 // This is needed for the polaris
 // styleguide to generate the props explorer
 interface Props extends IconProps {}
 
-export function Icon({source, color, backdrop, accessibilityLabel}: Props) {
+export function Icon({
+  source,
+  color: colorFromProps,
+  backdrop,
+  accessibilityLabel,
+}: Props) {
   const i18n = useI18n();
+  const {newDesignLanguage} = useFeatures();
+
+  const color =
+    colorFromProps == null && newDesignLanguage === true
+      ? 'base'
+      : colorFromProps;
+
+  let sourceType: 'function' | 'placeholder' | 'external';
+  if (typeof source === 'function') {
+    sourceType = 'function';
+  } else if (source === 'placeholder') {
+    sourceType = 'placeholder';
+  } else {
+    sourceType = 'external';
+  }
```

Comment:
This drys up the logic for determining the type of icon source, so we can reuse it to enable console warning below.